module/zfs/vdev_raidz.c

```diff
@@ -2036,22 +2082,30 @@ vdev_raidz_io_start(zio_t *zio)
 	raidz_map_t *rm;
 
 	if (vdrz->vd_logical_width != vdrz->vd_physical_width) {
-		/* XXX rangelock not needed after expansion completes */
-		zfs_locked_range_t *lr =
-		    zfs_rangelock_enter(&vdrz->vn_vre.vre_rangelock,
-		    zio->io_offset, zio->io_size, RL_READER);
-
-		rm = vdev_raidz_map_alloc_expanded(zio->io_abd,
-		    zio->io_size, zio->io_offset,
-		    tvd->vdev_ashift, vdrz->vd_physical_width,
-		    vdrz->vd_logical_width, vdrz->vd_nparity,
-		    vdrz->vn_vre.vre_offset_phys);
-		rm->rm_lr = lr;
-		/*
-		 * XXX If this is a write, will need to do additional
-		 * writes to locations that are already copied, but
-		 * not yet reflected in the on-disk format.
-		 */
+		uint64_t width = vdev_raidz_get_width(vdrz, zio->io_bp->blk_birth);
```

Comment:
since we're looking up based on logical birth time, I think this won't interact correctly with dedup blocks.  I think that `BP_PHYSICAL_BIRTH()` would work for both regular and dedup blocks.  BP_PHYSICAL_BIRTH() wouldn't work for blocks that were remapped due to device removal (see `remap_blkptr_cb()`).  However, device removal doesn't work with RAIDZ, so I think we might be OK.  It's definitely a bit tricky, though.